### PR TITLE
Fixes weird new List<Object>(size) syntax with new Object[size]

### PR DESCRIPTION
**Pull request by brianmfear**
Changes the output for new List<Object>(size); constructors to new Object[size].

This only does so safely; if we cannot determine if the change is safe, we fall back to List<Object>(size);. This is necessary because of the alternative new List<Object>(sourceListOrSet), which I'm not sure we can determine from the AST. No AST changes occur as a result of these changes.

Added a new set of tests to make sure we also handle nested list types, such as Map<String, List<String>> and Map<List<String>, List<String>>.

Fixed the code:

```apex
String[] g = new String[f];
```

In tests/new_object/NewObject.cls (newListTest), which is not valid if f is not an Integer variable; as a side effect, dynamic list sizes use the new List<Object>(size) constructor.

- [x] I’ve added tests to confirm my change works.
- [x] (If changing formatting output/API or CLI) I’ve added my changes to the `CHANGELOG.md` file in the `Unreleased` section.
- [x] I’ve read the [contributing guidelines](https://github.com/dangmai/prettier-plugin-apex/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -1271,32 +1271,36 @@ function handleNewSetLiteral(path, print) {
...
+  // We use List<Object>(param) otherwise.
+  // This should provide compatibility for all known types without knowing
+  // if the parameter is a variable (copy constructor) or literal size.
+
   const expressionDoc = path.call(print, "expr", "value");
   const parts = [];
+  const typePart = path.map(print, "types");
+  const hasLiteralNumberInitializer =
+    (typePart.group || (typePart.length && typePart[0].parts.length < 4)) &&
+    !Number.isNaN(parseInt(expressionDoc, 10));
```
Comment (dangmai) [inline src/printer.js:1290]: The AST does give us information about the literal type, so we don't need to parse this on our own. For example, this code will return whether the param type is Integer:

```
const node = path.getValue();
return node.expr && node.expr.type && node.expr.type['$'] === 'INTEGER';
```

File: src/printer.js [hunk 2]
```diff
@@ -1271,32 +1271,36 @@ function handleNewSetLiteral(path, print) {
...
-  // List<List<String>> data = new List<String>[]{};
-
-  // Other quirks:
-  // All of these are correct:
-  // Id[] ids = new Id[]{};
-  // Id[] ids = new Id[1];
-  // Id[] ids = new Id[anotherIds];
-  // But this is not correct:
-  // Id[] ids = new Id[1]{};
+  // We can delcare lists in the following ways:
```
Comment (dangmai) [inline src/printer.js:1274]: Small typo: declare instead of delcare

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 - Always break up TYPEOF SOQL queries ([issue](https://github.com/dangmai/prettier-plugin-apex/issues/244)).
 
+- Fixes weird `new List<Object>(0)` constructor with the more usual `new Object[0]` syntax when possible. Also adds new tests that cover edge cases.
```
Comment (dangmai) [inline CHANGELOG.md:5]: Feel free to credit yourself here, e.g. `(credit to @brianmfear)` at the end of this line. I'll add it after merge anyway :)

File: src/printer.js [hunk 4]
```diff
@@ -1271,32 +1271,36 @@ function handleNewSetLiteral(path, print) {
...
+  // We use Object[size] if a literal number is provided.
+  // We use List<Object>(param) otherwise.
+  // This should provide compatibility for all known types without knowing
+  // if the parameter is a variable (copy constructor) or literal size.
+
   const expressionDoc = path.call(print, "expr", "value");
   const parts = [];
+  const typePart = path.map(print, "types");
+  const hasLiteralNumberInitializer =
+    (typePart.group || (typePart.length && typePart[0].parts.length < 4)) &&
```
Comment (dangmai) [inline src/printer.js:1289]: I'm a little confused by this part. Why do we compare `parts` length to 4?

File: src/printer.js [hunk 5]
```diff
@@ -1271,32 +1271,36 @@ function handleNewSetLiteral(path, print) {
...
+  // new List<Object>(size);
+
+  // We use Object[size] if a literal number is provided.
+  // We use List<Object>(param) otherwise.
+  // This should provide compatibility for all known types without knowing
+  // if the parameter is a variable (copy constructor) or literal size.
+
   const expressionDoc = path.call(print, "expr", "value");
   const parts = [];
+  const typePart = path.map(print, "types");
```
Comment (dangmai) [inline src/printer.js:1287]: It looks like you are using this later to compare the values. Generally we only do `path.map(print, "blah")` if we intend to actually print it. If you need to get the value feel free to simply traverse the subtree:

```
const node = path.getValue();
// You can use `node.types` from here
```
We do it this way because calling `path.map(print)` adds some unnecessary stuffs to the `typePart` variable that is specific to printing.

**Review by dangmai — CHANGES_REQUESTED** [review]
Sorry for the long time it took me to get to this. I added some comments to your code, they're all pretty small so hopefully it won't take too long to fix.

File: src/printer.js (hunk 5 above)
Comment (brianmfear) [reply src/printer.js:1287]: Thanks. I walked in to this project with exactly zero experience, so I didn't know I could. I'll try to fix this.

File: src/printer.js (hunk 1 above)
Comment (brianmfear) [reply src/printer.js]: Thanks, I wasn't able to figure this one out with the apparently sparse documentation. I'm rather pro at browser debugging, not so much much CLI-based debugging.

**Review by brianmfear — COMMENTED** [review]
I have reviewed and amended changes, as requested.

**Comment by dangmai** [comment]
Merged in https://github.com/dangmai/prettier-plugin-apex/commit/68c281fc95a7de8ca68615ea71c463ba9af0a221
